src/wp-includes/class-wp-theme.php

```diff
@@ -1460,6 +1460,64 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
 		return false;
 	}
 
+	/**
+	 * Returns whether this theme is a block-based theme or not.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return bool
+	 */
+	public function is_block_based() {
+		$paths_to_index_block_template = array(
+			$this->get_file_path( '/block-templates/index.html' ),
+			$this->get_file_path( '/templates/index.html' ),
+		);
+
+		foreach ( $paths_to_index_block_template as $path_to_index_block_template ) {
+			if ( is_file( $path_to_index_block_template ) && is_readable( $path_to_index_block_template ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Retrieves the path of a file in the theme.
+	 *
+	 * Searches in the stylesheet directory before the template directory so themes
+	 * which inherit from a parent theme can just override one file.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param string $file Optional. File to search for in the stylesheet directory.
+	 * @return string The path of the file.
+	 */
+	public function get_file_path( $file = '' ) {
+		$file = ltrim( $file, '/' );
+
+		$stylesheet_directory = $this->get_stylesheet_directory();
+		$template_directory   = $this->get_template_directory();
+
+		if ( empty( $file ) ) {
+			$path = $stylesheet_directory;
+		} elseif ( file_exists( $stylesheet_directory . '/' . $file ) ) {
+			$path = $stylesheet_directory . '/' . $file;
+		} else {
+			$path = $template_directory . '/' . $file;
+		}
```

Comment:
What if the theme or requested file does not exist? For example:

- What happens if the theme doesn't exist? 
- What happens if the file does not exist in the child or parent?

In both cases, it flows through to the `else` and returns a path to a non-existent file.

Should this method check if the file exists before returning the path to it? There are checks in the `is_block_based()` method. But this method is public and could be invoked in a plugin, theme, or later in Core.

To avoid pushing the exists/readable checks downstream, maybe add them here:

```php

		// If the file does not exist, bail out and return an empty string.
		if ( ! is_readable( $path ) ) {
			return '';
		}
```

Reply:
@hellofromtonya 

> Should this method check if the file exists before returning the path to it?

The code of this method was copied over from the [`get_theme_file_path`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/link-template.php#L4485) function.
`WP_Theme::get_file_path` method is just an object-oriented version of that function (just as the `WP_Theme::is_block_based` method is an OO version of the [`wp_is_block_template_theme`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/theme.php#L4093) function).
The reason to introduce both `WP_Theme::get_file_path` and `WP_Theme::is_block_based` methods was that we have to be able to check if any given theme supports blocks (not just the current theme) on the `/wp-admin/themes.php` page.
In my opinion, these OO methods have to behave the same as their procedural counterparts (for backward compatibility).
If/when this PR gets merged into Core, I would like to create another trac issue to refactor  `get_theme_file_path` and `wp_is_block_template_theme` functions and make them call `WP_Theme::get_file_path` and  `WP_Theme::is_block_based` internally (to avoid code duplication).

> Should it check in this method to avoid pushing the exists/readable checks downstream?

We can add a second parameter to the `WP_Theme::get_file_path` method (equals false by default) if that is what we want. If that second parameter is true, we can return null if the file doesn't exist. Or ideally, we could add another public method to the `WP_Theme` class to implement this logic (to keep the number of arguments [as low as possible](https://github.com/jupeter/clean-code-php#function-arguments-2-or-fewer-ideally)).

Would you please let me know what your opinion on this matter is? Thank you.

Reply:
Thanks @anton-vlasenko for thoroughly explaining the reasoning for the code. Yes, I agree it makes sense (a) for these methods to be here in this class and (b) that this method uses a similar expected pattern of `get_theme_file_path()`. 

It could be interesting to add the option or another method that checks if the files exist. But I think for this PR what you have works, has solid reasons for them to be added, and fits what is needed for the attached issue. No need to add more than that at this point.